commands/migratelxc.go

```diff
@@ -0,0 +1,346 @@
+// LXC containers are not included in the map.
+func getLXCContainersFromState(st *state.State) (map[*state.Machine][]*state.Machine, error) {
+	// Collect LXC container machines by host.
+	hosts := make(map[string]*state.Machine)
+	byHost := make(map[*state.Machine][]*state.Machine)
+	machines, err := st.AllMachines()
+	if err != nil {
+		return nil, errors.Annotate(err, "getting machines")
+	}
+	for _, m := range machines {
+		if m.ContainerType() != "lxc" {
+			continue
+		}
+		parentId, _ := m.ParentId()
+		host, ok := hosts[parentId]
+		if !ok {
+			var err error
+			host, err = st.Machine(parentId)
+			if err != nil {
+				return nil, errors.Annotate(err, "getting host machine")
+			}
+			hosts[parentId] = host
+		}
+		byHost[host] = append(byHost[host], m)
+	}
+	return byHost, nil
+}
+
+// getLXDContainersFromMachines returns a map of host machines
+// to LXD containers contained within them. Hosts without LXD
+// containers are not included in the map.
+func getLXDContainersFromMachines(hosts []*state.Machine) (map[*state.Machine]map[string]*api.Container, error) {
+	var group errgroup.Group
+	lxdContainers := make([]map[string]*api.Container, len(hosts))
+	for i, host := range hosts {
+		i, host := i, host // copy for closure
+		group.Go(func() error {
+			containers, err := ListLXDContainers(host)
+			if err != nil {
+				return errors.Annotatef(err,
+					"listing LXD containers for host %q",
+					host.Id(),
+				)
+			}
+			lxdContainers[i] = containers
+			return nil
+		})
+	}
+	if err := group.Wait(); err != nil {
+		return nil, errors.Annotate(err, "listing LXD containers")
+	}
+	lxdByHost := make(map[*state.Machine]map[string]*api.Container)
+	for i, containers := range lxdContainers {
+		lxdByHost[hosts[i]] = containers
+	}
+	return lxdByHost, nil
+}
+
+type containerNames struct {
+	// oldName is the LXC container name used by Juju 1.25.
+	oldName string
+
+	// newName is the LXD container name used by Juju 2.x.
+	newName string
+}
+
+// getContainerNames returns the old (LXC) and new (LXD) container names
+// for the container entries recorded in state.
+func getContainerNames(st *state.State, byHost map[*state.Machine][]*state.Machine) (map[*state.Machine]containerNames, error) {
+	environUUID := st.EnvironUUID()
+	namespace, err := instance.NewNamespace(environUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	names := make(map[*state.Machine]containerNames)
+	for _, containers := range byHost {
+		for _, container := range containers {
+			// The recorded instance ID is the old
+			// LXC container name.
+			instanceId, err := container.InstanceId()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			oldName := string(instanceId)
+
+			// The new name, expected by Juju 2.x, uses a model
+			// UUID namespace.
+			newName, err := namespace.Hostname(container.Id())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+
+			names[container] = containerNames{
```

Comment:
This is much easier to think about than the dict mapping oldName to newName, nice one.